compiler/src/formatting/fmt.re

```diff
@@ -0,0 +1,3348 @@
+              )
+           ++ print_value_description(value_desc)
+         | PTopModule(provide_flag, module_decl) =>
+           (
+             switch (provide_flag) {
+             | NotProvided => empty
+             | Abstract => string("abstract ")
+             | Provided => string("provide ")
+             }
+           )
+           ++ print_comment_range(
+                ~allow_breaks=false,
+                ~trail=space,
+                enclosing_start_location(stmt.ptop_core_loc),
+                module_decl.pmod_loc,
+              )
+           ++ print_module_declaration(module_decl)
+         | PTopProvide(provide_items) =>
+           string("provide ")
+           ++ braces(
+                concat_map(
+                  ~lead=
+                    next =>
+                      print_comment_range(
+                        ~block_start=true,
+                        ~trail=space,
+                        enclosing_start_location(stmt.ptop_core_loc),
+                        switch (next) {
+                        | PProvideType({loc})
+                        | PProvideException({loc})
+                        | PProvideModule({loc})
+                        | PProvideValue({loc}) => loc
+                        },
+                      ),
+                  ~sep=
+                    (prev, next) =>
+                      comma
+                      ++ print_comment_range(
+                           ~none=breakable_space,
+                           ~lead=space,
+                           ~trail=breakable_space,
+                           switch (prev) {
+                           | PProvideType({loc})
+                           | PProvideException({loc})
+                           | PProvideModule({loc})
+                           | PProvideValue({loc}) => loc
+                           },
+                           switch (next) {
+                           | PProvideType({loc})
+                           | PProvideException({loc})
+                           | PProvideModule({loc})
+                           | PProvideValue({loc}) => loc
+                           },
+                         ),
+                  ~trail=
+                    prev =>
+                      trailing_comma
+                      ++ print_comment_range(
+                           ~block_end=true,
+                           ~lead=space,
+                           switch (prev) {
+                           | PProvideType({loc})
+                           | PProvideException({loc})
+                           | PProvideModule({loc})
+                           | PProvideValue({loc}) => loc
+                           },
+                           enclosing_end_location(stmt.ptop_core_loc),
+                         ),
+                  ~f=print_provide_item,
+                  provide_items,
+                ),
+              )
+         },
+       );
+  }
+
+  and print_comment_range =
```

Comment:
I need to spend a bunch more time understanding this function. I think it is trying to be too clever but I need to sit on it some more.

Reply:
#2038